kernel-config: Add config needed for waydroid

wentam opened the pull request:
My attempt at working out the minimal kernel configuration needed for waydroid with networking.

Adding this to #535 and normalizing kernel (with existing waydroid and firewall commits removed) results in a working setup.

A best-guess set of options was constructed through strace, reading netd source, and reading waydroid logcat. Was then able to whittle it down to the minimal set of options through trial-and-error.

Closes #437

wentam committed:
kernel-config: Add config needed for waydroid

samueldr commented on modules/kernel-config.nix:
```diff
@@ -83,6 +83,37 @@ in
+        ANDROID_BINDER_IPC = whenAtLeast "3.19" yes;
+        ANDROID_BINDERFS = whenAtLeast "5.0" yes;
+
+        # Needed for waydroid networking to function
+        NF_TABLES = whenAtLeast "3.13" yes;
+        NF_TABLES_IPV4 = mkMerge [ (whenBetween "3.13" "4.17" module) (whenAtLeast "4.17" yes) ];
+        NF_TABLES_IPV6 = mkMerge [ (whenBetween "3.13" "4.17" module) (whenAtLeast "4.17" yes) ];
+        NF_TABLES_INET = mkMerge [ (whenBetween "3.14" "4.17" module) (whenAtLeast "4.17" yes) ];
+        NFT_MASQ = whenAtLeast "3.18" module;
+        NFT_NAT = whenAtLeast "3.13" module;
+        IP_ADVANCED_ROUTER = whenAtLeast "2.5.45" yes;
+        IP_MULTIPLE_TABLES = whenAtLeast "2.5.45" yes;
+        IPV6_MULTIPLE_TABLES = whenAtLeast "2.6.19" yes;
+
+        # Needed for XfrmController
+        XFRM = whenAtLeast "2.6.0" yes;
+        XFRM_ALGO = whenAtLeast "3.5" module;
+        XFRM_USER = whenAtLeast "2.6.19" module;
+
+        # netd uses NFLOG
+        NETFILTER_NETLINK = whenAtLeast "2.6.25" yes;
+        NETFILTER_NETLINK_LOG = whenAtLeast "2.6.14" yes;
+        NETFILTER_XT_TARGET_NFLOG = whenAtLeast "2.6.20" module;
```
We don't need to gate anything older than the oldest in-use kernel version, `3.9` here.

Even `3.9` is getting too old for some things that e.g. systemd requires.

Also noting, but nothing to do for you, that we'll see in the future if it becomes problematic to re-define some symbol names like the networking bits here. I guess that as long as it resolves to the same answer it's a non-issue.

wentam replied:
> We don't need to gate anything older than the oldest in-use kernel version, `3.4` here.
> 
> Even `3.4` is getting too old for some things that e.g. systemd requires.

Removed unnecessary gates.

> Also noting, but nothing to do for you, that we'll see in the future if it becomes problematic to re-define some symbol names like the networking bits here. I guess that as long as it resolves to the same answer it's a non-issue.

I'd definitely like to see this being possible such that we can fully define the requirements for a given concern without any dependency. This way regardless of what happens with the firewall etc. waydroid should always function correctly as it's fully specified outside of any context. Just less fragile that way.

I can see a problematic corner-case with this, though: one place defines a config item as a module, another built-in. It would probably be best to make sure we have an assertion/warning if that occurs. Not important currently as we don't have modules, but thinking ahead here.

samueldr replied:
Yeah, it's more of a thought for *whenever that happens in the future*.

These configs aren't user-facing; they are for implementing a default kernel config. It is not expected that end-users will customize those, and if they will, that they know how to resolve the issues.

samueldr commented:
Converting to draft until kernel configs are normalized.

We probably want to handle this one after #536, right?

wentam commented:
> Converting to draft until kernel configs are normalized.

Should this PR contain normalized configs when/if this is ready to be merged? Slight risk as I obviously don't have all devices to test them.

> We probably want to handle this one after #536, right?

Makes sense. When testing it would be worth running a test with locally removed #536 though, to ensure this PR contains the correct coverage to get waydroid+network working without dependency (firewall has some overlap).

samueldr commented:
> Should this PR contain normalized configs when/if this is ready to be merged? Slight risk as I obviously don't have all devices to test them.

Yes. It is needed to:

 - Validate symbols resolve / apply
 - Kernels at least build with the new config

It is somewhat risky, there could be bugs in vendor kernels in unused bits of the kernel, but it is safe enough to expect that if it builds, it runs.

wentam commented:
Dropped CONFIG_ANDROID. [It has been removed](https://github.com/torvalds/linux/commit/1045a06724f322ed61f1ffb994427c7bdbe64647) and we really only care about the ANDROID_BINDER* stuff.

samueldr commented:
Maybe we want to add `versionOlder 6.1` (or whichever).

While it won't last for long for mainline-friendlier devices, some newer vendor kernels might benefit from it being enabled for waydroid? If that even can work on vendor kernels.

wentam commented:
> Maybe we want to add `versionOlder 6.1` (or whichever).
>
> While it won't last for long for mainline-friendlier devices, some newer vendor kernels might benefit from it being enabled for waydroid? If that even can work on vendor kernels.

CONFIG_ANDROID appears to be a [controversial feature](https://lwn.net/Articles/899743/) that can have some unexpected side-effects unrelated to waydroid.

For the same reasons that it was removed, it might not be wise to have this in our default global kernel configuration.

I also have not yet noticed anything broken in waydroid as a result of disabling CONFIG_ANDROID (on my 6.0 kernel). If any problems are noticed on pre-6.0 devices, then yes, it could be worth turning on for those devices. Otherwise I figure just don't take the risk.

samueldr commented:
Hm, I thought some of the options required for android (and thus waydroid) were gated behind `CONFIG_ANDROID`.

Reading the article, I see that maybe we actually want it. They are controversial on more desktop-style systems.

> At the core of the debate was the use of CONFIG_ANDROID as an indicator that the system will suspend and resume frequently

But this is a discussion point irrelevant for waydroid.

It's a good thing they're not just removing it, but making a more sensible implementation.

wentam commented:
I didn't really read into this very far. 'completely random config changes' mostly just sounded scary and my default is to enable less.

Given that `CONFIG_ANDROID` is probably the expected configuration for waydroid pre-6.0, I suppose it makes sense to turn it on.

Probably enable `CONFIG_PM_USERSPACE_AUTOSLEEP` for later kernels as well considering that mobile devices generally do suspend/resume frequently.

samueldr commented:
Yes, though I would think `PM_USERSPACE_AUTOSLEEP` can be handled independently from waydroid.

samueldr commented:
This is missing `PSI`. Or else `/dev/binder` goes away. (Will likely end-up sending the fix in this PR once I test more things out.)

Luflosi commented:
@wentam FYI: https://github.com/NixOS/mobile-nixos/pull/572 was merged.

Luflosi commented:
@wentam could you please resolve the conflict?

wentam commented:
* Rebased
* Added CONFIG_PSI
* Normalized all kernel configs (many kernels already contained these items and thus unchanged)

Ran a build of all kernels whose configs changed. All built fine.

samueldr committed:
asus-dumo: Normalize kernel config

samueldr committed:
mainline-chromeos-mt8183: Normalize kernel config

samueldr committed:
mainline-chromeos-sc7180: Normalize kernel config

samueldr committed:
sdm845-mainline: Normalize kernel config

samueldr committed:
motorola-potter: Normalize kernel config

samueldr committed:
pine64-pinephone: Normalize kernel config

samueldr committed:
pine64-pinetab: Normalize kernel config

samueldr committed:
additional-devices/uefi-x86_64-with-kernel: Normalize kernel config

samueldr commented:
The meat of the change (enabling waydroid) was left untouched.

I only dropped the kernel config normalization changes from that commit, and made them their own commits... And (obviously?) re-ran it. Since they're mechanical changes, there's no worries about it being done any differently.

samueldr commented:
Just built all kernels, ran one to confirm things are still fine (as they should still be).

Thank you, sorry it took a hot minute to get this merged.
